File: src/main/java/com/missfresh/util/HbaseUtil.java
package com.missfresh.util;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.HBaseConfiguration;
import org.apache.hadoop.hbase.HColumnDescriptor;
import org.apache.hadoop.hbase.HTableDescriptor;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.*;
import org.apache.hadoop.hbase.util.Bytes;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * @author wangzhihua
 * @date 2019-01-18 11:19
 */
public class HbaseUtil {
    public static Configuration conf = null;
    public static Connection connection = null;
    public static Admin admin = null;

    /**
     * @desc 取得连接
     */
    public  void setConf(String quorum, String port) {
        try {
            conf = HBaseConfiguration.create();
            conf.set("hbase.zookeeper.quorum", quorum);//zookeeper地址
            conf.set("hbase.zookeeper.property.clientPort", port);
            connection = ConnectionFactory.createConnection(conf);
            admin = connection.getAdmin();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * @desc 连关闭接
     */
    public void close() {
        try {
            if (connection != null) {
                connection.close();
            }
            if (admin != null) {
                admin.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * @desc 创建表
     */
    public  void createTable(String tableName, String columnFamily) {


        try {
            TableName tbName = TableName.valueOf(tableName);
            if (!admin.tableExists(tbName)) {
                HTableDescriptor hTableDescriptor = new HTableDescriptor(tbName);
                HColumnDescriptor hColumnDescriptor = new HColumnDescriptor(columnFamily);
                hTableDescriptor.addFamily(hColumnDescriptor);
                admin.createTable(hTableDescriptor);
            }
        } catch (IOException e) {
            e.printStackTrace();

        }


    }
    /**
     * 添加多条记录
     */
    public void addMoreRecord(String tableName, String family, String  qualifier, List < String > rowList, String value){
        Table table = null;
        try {
            table = connection.getTable(TableName.valueOf(tableName));

            List<Put> puts = new ArrayList<>();
            Put put = null;
            for (int i = 0; i < rowList.size(); i++) {
                put = new Put(Bytes.toBytes(rowList.get(i)));
                put.addColumn(Bytes.toBytes(family), Bytes.toBytes(qualifier), Bytes.toBytes(value));

                puts.add(put);
            }
            table.put(puts);

        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (table != null) {
                try {
                    table.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }
    //判断表是否存在
    public boolean isTableExist(String tableName) throws IOException {
        TableName tbName = null;
        try {
            tbName = TableName.valueOf(tableName);
            boolean flog=admin.tableExists(tbName);
            return  flog ;
        } catch (Exception e) {
            e.printStackTrace();
            if (connection != null) {
                connection.close();
            }
            if (admin != null) {
                admin.close();
            }
            return false;
        }


    }

    public static void main(String[] args) throws IOException {

        String family = "info";
        String tName = "rtlatform:abt_event_etl1";
        //Table table = null;
        String rowKey="40:10:21 62-30-9102155357163176703a176c7-8a02-b7c8-8538-a12bc4d94e63";
        System.out.println("0000");
        HbaseUtil hbaseUtil = new HbaseUtil();
        hbaseUtil.setConf("10.2.40.12,10.2.40.11,10.2.40.8", "2181");
        hbaseUtil.createTable(tName,"INFO");
       // hbaseUtil.isTableExist("rt_platform:AB_test");
        System.out.println("0000");
        System.out.println( hbaseUtil.isTableExist("rt_platform:AB_test"));
    }

}

File: src/main/java/com/missfresh/test/util/HbaseUtil.scala
package com.missfresh.test.util

import com.esotericsoftware.kryo.Kryo
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase._
import org.apache.hadoop.hbase.client._
import org.apache.hadoop.hbase.util.Bytes

import scala.collection.mutable

/**
  * Created by wangtaiyang on 2018/8/10. 
  */
object HbaseUtil {
  var conf: Configuration = _
  //线程池
  lazy val connection: Connection = ConnectionFactory.createConnection(conf)
  lazy val admin: Admin = connection.getAdmin

  /**
    * hbase conf
    *
    * @param quorum hbase的zk地址
    * @param port   zk端口2181
    * @return
    */
  def setConf(quorum: String, port: String): Unit = {
    val conf = HBaseConfiguration.create()
    conf.set("hbase.zookeeper.quorum", quorum)
    conf.set("hbase.zookeeper.property.clientPort", port)
    this.conf = conf
  }

  /**
    * 如果不存在就创建表
    *
    * @param tableName    命名空间：表名
    * @param columnFamily 列族
    */
  def createTable(tableName: String, columnFamily: String): Unit = {
    val tbName = TableName.valueOf(tableName)
    if (!admin.tableExists(tbName)) {
      val htableDescriptor = new HTableDescriptor(tbName)
      val hcolumnDescriptor = new HColumnDescriptor(columnFamily)
      htableDescriptor.addFamily(hcolumnDescriptor)
      admin.createTable(htableDescriptor)
    }
  }

  def hbaseScan(tableName: String): ResultScanner = {
    val scan = new Scan()
    val table = connection.getTable(TableName.valueOf(tableName))
    table.getScanner(scan)
    //    val scanner: CellScanner = rs.next().cellScanner()
  }

  /**
    * 获取hbase单元格内容
    *
    * @param tableName 命名空间：表名
    * @param rowKey    rowkey
    * @return 返回单元格组成的List
    */
  def getCell(tableName: String, rowKey: String): mutable.Buffer[Cell] = {
    val get = new Get(Bytes.toBytes(rowKey))
    val table = connection.getTable(TableName.valueOf(tableName))
    val result: Result = table.get(get)
    import scala.collection.JavaConverters._
    result.listCells().asScala
  }

  /**
    * 单条插入
    *
    * @param tableName 命名空间：表名
    * @param rowKey    rowkey
    * @param family    列族
    * @param qualifier column列
    * @param value     列值
    */
  def singlePut(tableName: String, rowKey: String, family: String, qualifier: String, value: String): Unit = {
    //向表中插入数据//向表中插入数据
    //a.单个插入
    val put: Put = new Put(Bytes.toBytes(rowKey)) //参数是行健row01
    put.addColumn(Bytes.toBytes(family), Bytes.toBytes(qualifier), Bytes.toBytes(value))

    //获得表对象
    val table: Table = connection.getTable(TableName.valueOf(tableName))

    table.put(put)
    table.close()

  }

  /**
    * 删除数据
    *
    * @param tbName 表名
    * @param row    rowkey
    */
  def deleteByRow(tbName: String, row: String): Unit = {
    val delete = new Delete(Bytes.toBytes(row))
    //    delete.addColumn(Bytes.toBytes("fm2"), Bytes.toBytes("col2"))
    val table = connection.getTable(TableName.valueOf(tbName))
    table.delete(delete)
  }

  def close(): Unit = {
    admin.close()
    connection.close()
  }

  def main(args: Array[String]): Unit = {
    setConf("localhost", "2181")
    try{
      createTable("kafka_offset:topic_offset_range", "info")
      singlePut("kafka_offset:topic_offset_range", "gid_topic_name", "info", "partition0", "200")
      singlePut("kafka_offset:topic_offset_range", "gid_topic_name", "info", "partition1", "300")
      singlePut("kafka_offset:topic_offset_range", "gid_topic_name", "info", "partition2", "100")

    }catch {
      case e:Exception=>println("----")
    }

    this.close()


  }
}
